src/Codec/Xlsx/Parser/Stream.hs

```diff
@@ -261,61 +263,72 @@ parseSharedStrings = \case
+      sharedStrsSel <- Zip.mkEntrySelector "xl/sharedStrings.xml"
+      hasSharedStrs <- Zip.doesEntryExist sharedStrsSel
+      if not hasSharedStrs
         then pure mempty
         else do
           let state0 = initialSharedStrings
-          byteSrc <- liftZip $ Zip.getEntrySource sharedStrsSel
-          st <- runExpat state0 byteSrc $ \evs -> forM_ evs $ \ev -> do
+          byteSrc <- Zip.getEntrySource sharedStrsSel
+          st <- liftIO $ runExpat state0 byteSrc $ \evs -> forM_ evs $ \ev -> do
             mTxt <- parseSharedStrings ev
             for_ mTxt $ \txt ->
               ss_list %= (`DL.snoc` txt)
           pure $ V.fromList $ DL.toList $ _ss_list st
-      liftIO $ writeIORef sharedStringsRef $ Just sharedStrs
-      pure sharedStrs
+
+{-# SCC getOrParseSharedStringss #-}
+getOrParseSharedStringss :: XlsxM (V.Vector Text)
+getOrParseSharedStringss = runMemoized =<< asks _xs_shared_strings
+
+readWorkbookInfo :: Zip.ZipArchive WorkbookInfo
+readWorkbookInfo = do
+   sel <- Zip.mkEntrySelector "xl/workbook.xml"
+   src <- Zip.getEntrySource sel
+   sheets <- liftIO $ runExpat [] src $ \evs -> forM_ evs $ \case
+     StartElement ("sheet" :: ByteString) attrs -> do
+       nm <- lookupBy "name" attrs
+       sheetId <- lookupBy "sheetId" attrs
+       rId <- lookupBy "r:id" attrs
+       sheetNum <- either (throwM . ParseDecimalError sheetId) pure $ eitherDecimal sheetId
+       modify' (SheetInfo nm (RefId rId) sheetNum :)
+     _ -> pure ()
+   pure $ WorkbookInfo sheets
+
+lookupBy :: MonadThrow m => ByteString -> [(ByteString, Text)] -> m Text
+lookupBy fields attrs = maybe (throwM $ LookupError attrs fields) pure $ lookup fields attrs
 
 -- | Returns information about the workbook, found in
 -- xl/workbook.xml. The result is cached so the XML will only be
 -- decompressed and parsed once inside a larger XlsxM action.
 getWorkbookInfo :: XlsxM WorkbookInfo
-getWorkbookInfo = do
-  ref <- asks _xs_workbook_info
-  mInfo <- liftIO $ readIORef ref
-  case mInfo of
-    Just info -> pure info
-    Nothing -> do
-      sel <- liftZip $ Zip.mkEntrySelector "xl/workbook.xml"
-      src <- liftZip $ Zip.getEntrySource sel
-      sheets <- runExpat [] src $ \evs -> forM_ evs $ \case
-        StartElement ("sheet" :: ByteString) attrs -> do
-          nm <- maybe (throwM MalformedWorkbook) pure $ lookup ("name" :: ByteString) attrs
-          sheetId <- maybe (throwM MalformedWorkbook) pure $ lookup "sheetId" attrs
-          rId <- maybe (throwM MalformedWorkbook) pure $ lookup "r:id" attrs
-          sheetNum <- either (const $ throwM MalformedWorkbook) pure $ eitherDecimal sheetId
-          modify' (SheetInfo nm (RefId rId) sheetNum :)
-        _ -> pure ()
-      let info = WorkbookInfo sheets
-      liftIO $ writeIORef ref $ Just info
-      pure info
+getWorkbookInfo = runMemoized =<< asks _xs_workbook_info
+
+readWorkbookRelatinoships :: Zip.ZipArchive Relationships
+readWorkbookRelatinoships = do
```

Comment:
a typo

Reply:
I fixed it, this function wasn't exposed so no problem.